x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,332 @@
+        }
+        if (pinnedDocs != null) {
+            out.writeBoolean(true);
+            out.writeList(pinnedDocs);
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    public List<String> rulesetIds() {
+        return rulesetIds;
+    }
+
+    public Map<String, Object> matchCriteria() {
+        return matchCriteria;
+    }
+
+    public QueryBuilder organicQuery() {
+        return organicQuery;
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(ORGANIC_QUERY_FIELD.getPreferredName(), organicQuery);
+        builder.startObject(MATCH_CRITERIA_FIELD.getPreferredName());
+        builder.mapContents(matchCriteria);
+        builder.endObject();
+        builder.startArray(RULESET_IDS_FIELD.getPreferredName());
+        for (String rulesetId : rulesetIds) {
+            builder.value(rulesetId);
+        }
+        builder.endArray();
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
+            throw new IllegalArgumentException("Can't have both pinned ids and pinned docs");
+        }
+
+        if (pinnedIds != null && pinnedIds.isEmpty() == false) {
+            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedIds.toArray(new String[0]));
+            return pinnedQueryBuilder.toQuery(context);
+        } else if (pinnedDocs != null && pinnedDocs.isEmpty() == false) {
+            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedDocs.toArray(new Item[0]));
+            return pinnedQueryBuilder.toQuery(context);
+        } else {
+            return organicQuery.toQuery(context);
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        if (pinnedIds != null || pinnedDocs != null) {
+            return this;
+        } else if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
+            List<String> identifiedPinnedIds = pinnedIdsSupplier != null ? pinnedIdsSupplier.get() : null;
+            List<Item> identifiedPinnedDocs = pinnedDocsSupplier != null ? pinnedDocsSupplier.get() : null;
+            if (identifiedPinnedIds == null && identifiedPinnedDocs == null) {
+                return this; // not executed yet
+            } else {
+                return new RuleQueryBuilder(organicQuery, matchCriteria, rulesetIds, identifiedPinnedIds, identifiedPinnedDocs, null, null);
+            }
+        }
+
+        // Identify matching rules and apply them as applicable
+        String rulesetId = rulesetIds.get(0);
```

Comment:
Note: I'm currently verifying with @aznick if we can trim the scope down to only allowing a single `ruleset_id` in the request rather than supporting a list of ruleset IDs.